File: py_example_service/py_example_service/example_server.py
#! /usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from custom_interfaces.srv import TurningControl
import time



# 노드 생성하기
class ExampleServerNode(Node):
        # 생성자
        def __init__(self):
                super().__init__("example_server_node")

                # Service Server 생성하기
                self.server = self.create_service(TurningControl, "example_service", self.run_callback)

                # publisher 생성하기
                self.publisher = self.create_publisher(Twist, "skidbot/cmd_vel", 10)

                self.get_logger().info("example_server start")

        # move_robot
        def move_robot(self, duration=1, linear_vel_x=0.0, angular_vel_z=0.0):
                self.get_logger().info("move_robot")

                twist_msg = Twist()
                twist_msg.linear.x = linear_vel_x
                twist_msg.angular.z = angular_vel_z

                start_time = self.get_clock().now().to_msg().sec
                end_time = self.get_clock().now().to_msg().sec

                while (end_time - start_time) < duration:
                        self.publisher.publish(twist_msg)
                        end_time = self.get_clock().now().to_msg().sec
                        time.sleep(0.01)

        # stop_robot
        def stop_robot(self):
                self.get_logger().info("stop_robot")

                twist_msg = Twist()
                twist_msg.linear.x = 0.0
                twist_msg.angular.z = 0.0

                self.publisher.publish(twist_msg)

        # callback
        def run_callback(self, request, response):
                self.get_logger().info("run_callback")

                # control robot
                self.move_robot(request.time_duration, request.linear_vel_x, request.angular_vel_z)
                self.stop_robot()

                response.success = True
                return response

# main()
def main(args=None):
        # 초기화하기
        rclpy.init(args=args)

        # Node 생성하기
        example_server_node = ExampleServerNode()

        # 실행하기
        rclpy.spin(example_server_node)

        # 종료하기
        example_server_node.destroy_node()
        rclpy.shutdown()

if __name__ == "__main__":
        main()

File: cpp_example_service/src/example_server.cpp
#include <unistd.h>

#include "rclcpp/rclcpp.hpp"
#include "cpp_example_service/srv/turning_control.hpp"
#include "geometry_msgs/msg/twist.hpp"



// Node 클래스 정의하기
class ExampleServerNode : public rclcpp::Node
{
private:
        rclcpp::Service<cpp_example_service::srv::TurningControl>::SharedPtr m_service;

        rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr m_publisher;

public:
        // 생성자
        ExampleServerNode() : rclcpp::Node("example_server_node")
        {
                RCLCPP_INFO(get_logger(), "ExampleServerNode start");

                // create publisher : skidbot/cmd_vel
                m_publisher = create_publisher<geometry_msgs::msg::Twist>("skidbot/cmd_vel", 10);

                // create service : turn_robot
                m_service = create_service<cpp_example_service::srv::TurningControl>(
                        "turn_robot", std::bind(&ExampleServerNode::response_callback, this, std::placeholders::_1, std::placeholders::_2));
        }

        // response callback
        void response_callback(std::shared_ptr<cpp_example_service::srv::TurningControl::Request> request, std::shared_ptr<cpp_example_service::srv::TurningControl::Response> response)
        {
                RCLCPP_INFO(get_logger(), "response_callback");

                auto start_time = now();
                auto end_time = now();
                auto duration_time = request->time_duration * 1e9;

                while ((end_time - start_time).nanoseconds() < duration_time)
                {
                        move_robot(request->linear_vel_x, request->angular_vel_z);
                        end_time = now();
                        usleep(100000);
                }
                
                stop_robot();

                response->success = true;
        }

        // move robot
        void move_robot(const float linear_vel_x, const float angular_vel_z)
        {
                RCLCPP_INFO(get_logger(), "move_robot");
                geometry_msgs::msg::Twist twist_msg;

                twist_msg.linear.x = linear_vel_x;
                twist_msg.angular.z = angular_vel_z;

                m_publisher->publish(twist_msg);
        }

        // stop robot
        void stop_robot(void)
        {
                RCLCPP_INFO(get_logger(), "stop robot");
                geometry_msgs::msg::Twist twist_msg;

                twist_msg.linear.x = 0.0;
                twist_msg.angular.z = 0.0;

                m_publisher->publish(twist_msg);
        }
};



/*
 *
 */
int main(int argc, char **argv)
{
        // 초기화하기
        rclcpp::init(argc, argv);

        // 노드 생성하기
        auto example_server_node = std::make_shared<ExampleServerNode>();

        // 대기하기
        rclcpp::spin(example_server_node);

        // 종료하기
        rclcpp::shutdown();

        return 0;
}

